NIS-5220 Implement Authorization API

aleh-kasiankou-innowise opened the pull request:
Ready:
NIS-5229 - US-1 Sign up
NIS-5230 - US-2 Sign in
NIS-5231 - US-3 Sign out

Aleh-Kasiankou committed:
add dummy db context

Aleh-Kasiankou committed:
add docker-compose

Aleh-Kasiankou committed:
add basic patient registration

Aleh-Kasiankou committed:
add password max-length validation

+ remove unused ILogger dependency from AuthenticationController

Aleh-Kasiankou committed:
fix authentication scheme + add swagger jwt support

Aleh-Kasiankou committed:
move service configuration to DependencyInjection assembly

Aleh-Kasiankou committed:
fix assembly reference + add roles

Aleh-Kasiankou committed:
add roles on model creating

Aleh-Kasiankou committed:
add refresh token

Aleh-Kasiankou committed:
add integration tests

Aleh-Kasiankou committed:
add patient sign-in

Aleh-Kasiankou committed:
fix email uniqueness validation

Aleh-Kasiankou committed:
add sign-in integration tests

Aleh-Kasiankou committed:
add user sign-out

Aleh-Kasiankou committed:
add integration tests

+ add token revoke on login fail
+ add constants with controller/endpoints routes

aleh-kasiankou-innowise committed:
Merge pull request #1 from aleh-kasiankou-innowise/feature/NIS-5229

Feature/NIS-5229

aleh-kasiankou-innowise committed:
Merge pull request #2 from aleh-kasiankou-innowise/feature/NIS-5230

Feature/NIS-5230

aleh-kasiankou-innowise committed:
Merge pull request #3 from aleh-kasiankou-innowise/feature/NIS-5231

Feature/NIS-5231

aleh-kasiankou-innowise committed:
fix Dto assembly name case

aleh-kasiankou-innowise committed:
separate integration tests to groups by testing object

aleh-kasiankou-innowise committed:
increase jwt token lifetime + replace config for testing

-- token lifetime has been increased to 15 minutes
-- token lifetime for integration testing has been decreased to 5 seconds
-- token lifetime unit of measurement has been switched from minutes to seconds

aleh-kasiankou-innowise committed:
fix root namespace + add xml comments

aleh-kasiankou-innowise committed:
add email confirmation on registration

aleh-kasiankou-innowise committed:
Merge remote-tracking branch 'origin/development' into feature/NIS-5230

# Conflicts:
#	Innowise.Clinic.Auth/Controllers/AuthenticationController.cs
#	Innowise.Clinic.Auth/Program.cs

aleh-kasiankou-innowise committed:
resolve merge conflicts with develop

aleh-kasiankou-innowise committed:
Merge pull request #4 from aleh-kasiankou-innowise/refactoring/xml_comments_and_tests

Refactoring/xml_comments_and_tests

aleh-kasiankou-innowise committed:
Merge remote-tracking branch 'origin/development' into feature/NIS-5230

# Conflicts:
#	Innowise.Clinic.Auth.Api/Controllers/AuthenticationController.cs
#	Innowise.Clinic.Auth.Api/Innowise.Clinic.Auth.Api.csproj
#	docker-compose/docker-compose.yml

aleh-kasiankou-innowise committed:
fix integration tests

-- add smtp service docker container for each test group
-- fix SmtpData.cs property naming

aleh-kasiankou-innowise committed:
add xml comments for email confirmation email

aleh-kasiankou-innowise committed:
Merge pull request #5 from aleh-kasiankou-innowise/feature/NIS-5230

Feature/NIS-5230

sdechko1 commented on Innowise.Clinic.Auth.Api/Controllers/AuthenticationController.cs:
```diff
@@ -0,0 +1,252 @@
+///  and revokes tokens for whose willing to log out.
+/// </summary>
+[ApiController]
+[Route(ControllerRoutes.AuthenticationControllerRoute)]
```
For this attribute you do not need to manually create route every time. You can pass `"api/[controller]"` to attribute constructor and will pick Controller's name with omitted `"Controller"` at the end.

sdechko1 commented on Innowise.Clinic.Auth.Api/Program.cs:
```diff
@@ -0,0 +1,64 @@
+
+app.Run();
+
+namespace Innowise.Clinic.Auth.Api
+{
+    public partial class Program
+    {
+    }
+}
```
Why do we need this ?

aleh-kasiankou-innowise replied:
This is necessary for running integration tests:
https://learn.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-6.0#basic-tests-with-the-default-webapplicationfactory-1

sdechko1 commented on Innowise.Clinic.Auth.Api/Controllers/AuthenticationController.cs:
```diff
@@ -0,0 +1,252 @@
+
+                var emailConfirmationToken = await _userManager.GenerateEmailConfirmationTokenAsync(user);
+
+                byte[] tokenGeneratedBytes = Encoding.UTF8.GetBytes(emailConfirmationToken);
```
Why type here is written explicitly ?

sdechko1 commented on Innowise.Clinic.Auth.Api/Controllers/AuthenticationController.cs:
```diff
@@ -0,0 +1,252 @@
+    [HttpPost(EndpointRoutes.SignUpEndpointRoute)]
+    [ProducesResponseType(typeof(AuthTokenPairDto), 200)]
+    [ProducesResponseType(typeof(void), 400)]
+    public async Task<ActionResult<AuthTokenPairDto>> RegisterPatient(PatientCredentialsDto patientCredentials)
```
Try to extract some code to service because controller method is too big.

sdechko1 commented on Innowise.Clinic.Auth.Api/Controllers/AuthenticationController.cs:
```diff
@@ -0,0 +1,252 @@
+[ApiController]
+[Route(ControllerRoutes.AuthenticationControllerRoute)]
+[Produces("application/json")]
+public class AuthenticationController : ControllerBase
```
This controller contains too much functionality that can be extracted to services. For example to `TokenGenerator`

sdechko1 commented on Innowise.Clinic.Auth.Api/Controllers/AuthenticationController.cs:
```diff
@@ -0,0 +1,252 @@
+    public async Task<ActionResult<AuthTokenPairDto>> RegisterPatient(PatientCredentialsDto patientCredentials)
+    {
+        var userExists = await _userManager.FindByEmailAsync(patientCredentials.Email);
+        if (userExists == null)
```
This if statement is too nested. Try inverse if condition clause.

sdechko1 commented on Innowise.Clinic.Auth.Api/Controllers/AuthenticationController.cs:
```diff
@@ -0,0 +1,252 @@
+        catch (ApplicationException e)
+        {
+            return Unauthorized(e.Message);
+        }
```
If you get Exception that is not child of ApplicationException your application will crash. Add additional catch clause for all exceptions. Or add global error handling by using custom middleware.

sdechko1 commented on Innowise.Clinic.Auth.Api/Controllers/AuthenticationController.cs:
```diff
@@ -0,0 +1,252 @@
+            var tokenId = tokens.GetRefreshTokenId();
+            await _tokenRevoker.RevokeTokenAsync(tokenId);
+        }
+
```
Remove this unneccessary blank lines.

sdechko1 commented on Innowise.Clinic.Auth.Api/Controllers/AuthenticationController.cs:
```diff
@@ -0,0 +1,252 @@
+            var principal = await validator.ValidateTokenPairAndExtractPrincipal(tokens, true);
+            return _tokenGenerator.GenerateJwtToken(principal);
+        }
+        catch (Exception e)
```
you need to make exception handling more type specific because you don't want a user to be unauthorized if app throws for example NullReferenceException.

sdechko1 commented on Innowise.Clinic.Auth.Api/Controllers/ValidationController.cs:
```diff
@@ -0,0 +1,42 @@
+    {
+        var validationSucceeded = await _emailValidator.ValidateEmailAsync(email);
+
+        return validationSucceeded? Ok() : BadRequest(EmailIsRegisteredMessage);
```
it would be much clearer if it will be like 
```
        return validationSucceeded
            ? Ok()
            : BadRequest(EmailIsRegisteredMessage);
```

sdechko1 commented on Innowise.Clinic.Auth.Api/Controllers/ValidationController.cs:
```diff
@@ -0,0 +1,42 @@
+
+    public async Task<IActionResult> CheckEmailUniqueness([FromRoute] string email)
+    {
+        var validationSucceeded = await _emailValidator.ValidateEmailAsync(email);
```
Booleans supposed to start with `is` like `isValidationSucceeded`

sdechko1 commented on Innowise.Clinic.Auth.Constants/ApiMessage.cs:
```diff
@@ -0,0 +1,6 @@
+
+public static class ApiMessages
+{
+    public const string FailedLoginMessage = "Either an email or a password is incorrect";
```
You use this string only once in the app so why do we need separate class for it ?

sdechko1 commented on Innowise.Clinic.Auth.Constants/ControllerRoutes.cs:
```diff
@@ -0,0 +1,9 @@
+    public const string AuthenticationControllerRoute = "auth";
+    public const string ValidationControllerRoute = "auth/validation";
+
+
```
Remove this blank lines.

sdechko1 commented on Innowise.Clinic.Auth.Constants/EndpointRoutes.cs:
```diff
@@ -0,0 +1,9 @@
+namespace Innowise.Clinic.Auth.Constants;
+
+public static class EndpointRoutes
```
As you use this contants once per each you don't need them as separate static class.

sdechko1 commented on Innowise.Clinic.Auth.Dto/AuthTokenPairDto.cs:
```diff
@@ -0,0 +1,47 @@
+/// </summary>
+public class AuthTokenPairDto
+{
+#pragma warning disable CS1591
```
Do not use pragma to suppress this kind of warnings.

sdechko1 commented on Innowise.Clinic.Auth.Dto/AuthTokenPairDto.cs:
```diff
@@ -0,0 +1,47 @@
+        RefreshToken = refreshToken;
+    }
+
+    private AuthTokenPairDto()
+    {
+    }
```
Why do we need this ?

sdechko1 commented on Innowise.Clinic.Auth.Dto/AuthTokenPairDto.cs:
```diff
@@ -0,0 +1,47 @@
+    /// JQYXRpZW50IiwiZXhwIjoxNjc1NjI2MTA1LCJpc3MiOiJo
+    /// dHRwOi8vbG9jYWxob3N0OjI1NTcwIn0.acLuMB4OHsW4at
+    /// 8Y9hmXNwopYDxbcZ8TSiAvWVgf7L8</example>
+    public string SecurityToken { get; set; }
```
set accessors are not used so we dont need them.

sdechko1 commented on Innowise.Clinic.Auth.Dto/PatientCredentialsDto.cs:
```diff
@@ -0,0 +1,25 @@
+    /// <example>patient@gmail.com</example>
+    [EmailAddress]
+    [Required(ErrorMessage = "Email is required")]
+    public string Email { get; set; }
```
replace `set` by `init` or use `Record`.

sdechko1 commented on Innowise.Clinic.Auth.Extensions/ConfigurationManager.cs:
```diff
@@ -0,0 +1,107 @@
+    public static IServiceCollection ConfigureIdentity(this IServiceCollection services)
+    {
+        services.AddIdentity<IdentityUser<Guid>, IdentityRole<Guid>>(options =>
+            {
+                options.Password.RequireDigit = false;
+                options.Password.RequiredLength = 6;
+                options.Password.RequireLowercase = false;
+                options.Password.RequireNonAlphanumeric = false;
+                options.Password.RequireUppercase = false;
+            })
```
This should be in configuration file.

sdechko1 commented on Innowise.Clinic.Auth.Extensions/ConfigurationManager.cs:
```diff
@@ -0,0 +1,107 @@
+            options.DefaultAuthenticateScheme = JwtBearerDefaults.AuthenticationScheme;
+            options.DefaultChallengeScheme = JwtBearerDefaults.AuthenticationScheme;
+        }).AddJwtBearer(options =>
+        {
+            options.SaveToken = true;
+            options.RequireHttpsMetadata = false;
+            options.TokenValidationParameters = new TokenValidationParameters()
+            {
+                ValidateIssuer = true,
+                ValidIssuer = jwtData["ValidIssuer"],
+                ValidateAudience = false,
+                ValidateIssuerSigningKey = true,
+                IssuerSigningKey =
+                    new SymmetricSecurityKey(Encoding.UTF8.GetBytes(jwtData["Key"]))
+            };
```
Settings for configuration JWT should be in config file.

sdechko1 commented on Innowise.Clinic.Auth.Extensions/ConfigurationManager.cs:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    public static IServiceCollection ConfigureJwtAuthentication(this IServiceCollection services,
+        IConfigurationSection jwtData)
```
jwtData looks like it is token itself so rename it ito smth like `jwtConfig `or `jwtSettings`

sdechko1 commented on Innowise.Clinic.Auth.Extensions/ConfigurationManager.cs:
```diff
@@ -0,0 +1,107 @@
+        });
+    }
+
+    public static void ConfigureCustomValidators(this IServiceCollection services)
```
Why return type is void? What I should do if I want to chain configurations?

sdechko1 commented on Innowise.Clinic.Auth.Extensions/Exceptions/PrincipalLacksTokenIdException.cs:
```diff
@@ -0,0 +1,11 @@
+{
+    private const string DefaultMessage = "The principal lacks TokenID";
+    
+    public PrincipalLacksTokenIdException() : base(DefaultMessage)
```
What if I want to pass my custom message to ctor ?

sdechko1 commented on Innowise.Clinic.Auth.IntegrationTesting/IntegrationTestingWebApplicationFactory.cs:
```diff
@@ -0,0 +1,134 @@
+    private readonly TestcontainersContainer _dbContainer;
+    private readonly TestcontainersContainer _mailContainer;
+
+    private const string ContainerHost = "localhost";
```
const is mixed with readonly variables so you need to separate them.

sdechko1 commented on Innowise.Clinic.Auth.IntegrationTesting/SignInTests.cs:
```diff
@@ -0,0 +1,109 @@
+
+        // Assert
+
+        Assert.True(response.IsSuccessStatusCode);
```
For testing we have more modern library "FluentAssertion"

sdechko1 commented on Innowise.Clinic.Auth.IntegrationTesting/TestHelper.cs:
```diff
@@ -0,0 +1,66 @@
+public static class TestHelper
+{
+    
+    internal const string SignUpEndpointUri =
+        ControllerRoutes.AuthenticationControllerRoute + "/" + EndpointRoutes.SignUpEndpointRoute;
+
+    internal const string RefreshTokenEndpointUri =
+        ControllerRoutes.AuthenticationControllerRoute + "/" + EndpointRoutes.RefreshTokenEndpointRoute;
+
+    internal const string SignInEndpointUri = ControllerRoutes.AuthenticationControllerRoute + "/" +
+                                              EndpointRoutes.SignInEndpointRoute;
+
+    internal const string SignOutEndpointUri = ControllerRoutes.AuthenticationControllerRoute + "/" +
+                                               EndpointRoutes.SignOutEndpointRoute;
+
+    
+    internal static int _uniqueNumber = 0;
+
+    internal static int UniqueNumber => _uniqueNumber++;
```
Constants in tests is bad practice because if we change value of constant tests will not fail but all frontends will.

sdechko1 commented on Innowise.Clinic.Auth.IntegrationTesting/TestHelper.cs:
```diff
@@ -0,0 +1,66 @@
+
+namespace Innowise.Clinic.Auth.IntegrationTesting;
+
+public static class TestHelper
```
All fields and methds are internal so why class is public ?

sdechko1 commented on Innowise.Clinic.Auth.IntegrationTesting/TokenGenerationTests.cs:
```diff
@@ -0,0 +1,286 @@
+            await _httpClient.PostAsJsonAsync(TestHelper.SignUpEndpointUri, userRegistrationDataWithRegisteredEmail);
+        var generatedTokens = await response.Content.ReadFromJsonAsync<AuthTokenPairDto>();
+
+        await Task.Delay(_jwtData.Value.TokenValidityInSeconds * 1000);
```
if you configure validity to smth like 60 minutes the test will be executing forever :).

sdechko1 commented on Innowise.Clinic.Auth.Jwt/RefreshTokenRevoker.cs:
```diff
@@ -0,0 +1,75 @@
+
+    public async Task RevokeTokenAsync(Guid tokenId)
+    {
+        var token = await _dbContext.RefreshTokens.SingleOrDefaultAsync(x => x.TokenId == tokenId);
```
For those purposes FirstOrDefault is better.

sdechko1 commented on Innowise.Clinic.Auth.Jwt/RefreshTokenRevoker.cs:
```diff
@@ -0,0 +1,75 @@
+        string tableName = modelEntityType?.GetSchemaQualifiedTableName() ??
+                           throw new InvalidOperationException("The required table doesn't exist");
+
+        await _dbContext.Database.ExecuteSqlRawAsync($"TRUNCATE TABLE {tableName};");
```
Why here you use raw sql besides LINQ ?

sdechko1 commented on Innowise.Clinic.Auth.Jwt/TokenGenerator.cs:
```diff
@@ -0,0 +1,73 @@
+
+namespace Innowise.Clinic.Auth.Jwt;
+
+public class TokenGenerator : ITokenGenerator
```
Generators are associated with static classes so rename it to smth like `TokenService`.

sdechko1 commented on Innowise.Clinic.Auth.Api/Controllers/AuthenticationController.cs:
```diff
@@ -0,0 +1,252 @@
+    {
+        try
+        {
+            await validator.ValidateTokenPairAndExtractPrincipal(tokens, false);
```
What is an aim for this call ?

sdechko1 commented on Innowise.Clinic.Auth.Jwt/TokenValidator.cs:
```diff
@@ -0,0 +1,97 @@
+    private ClaimsPrincipal ValidateTokenAndReturnPrinciple(string token, bool validateExpirationDate,
+        out SecurityToken validatedToken)
+    {
+        var jwtTokenValidationParameters = new TokenValidationParameters
+        {
+            ValidateIssuer = true,
+            ValidIssuer = _jwtConfiguration.Value.ValidIssuer,
+            ValidateIssuerSigningKey = true,
+            ValidateAudience = false,
+            IssuerSigningKey = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(_jwtConfiguration.Value.Key)),
+            ValidateLifetime = validateExpirationDate
+        };
```
This also need to be in config files.

sdechko1 commented on Innowise.Clinic.Auth.Mail/EmailSender.cs:
```diff
@@ -0,0 +1,42 @@
+
+public class EmailSender : IEmailHandler
+{
+    private readonly IOptions<SmtpData> _smtpData;
```
Why class is named sender when interface is Handler ?

sdechko1 commented on Innowise.Clinic.Auth.Mail/EmailSender.cs:
```diff
@@ -0,0 +1,42 @@
+{
+    private readonly IOptions<SmtpData> _smtpData;
+
+    public EmailSender(IOptions<SmtpData> emailData)
```
emailConfiguration and not emailData

sdechko1 commented on Innowise.Clinic.Auth.Mail/EmailSender.cs:
```diff
@@ -0,0 +1,42 @@
+        _smtpData = emailData;
+    }
+
+    public void SendMessage(string mailRecipientMailAddress, string mailSubject, string mailBody)
```
I think this method should be async since it works with net.

sdechko1 commented on Innowise.Clinic.Auth.Mail/SmtpData.cs:
```diff
@@ -0,0 +1,9 @@
+namespace Innowise.Clinic.Auth.Mail;
+
+public class SmtpData
```
SmtpOptions and not SmtpData

sdechko1 commented on Innowise.Clinic.Auth.Persistence/UserRoleConfiguration.cs:
```diff
@@ -0,0 +1,34 @@
+                Name = UserRoles.Receptionist,
+                NormalizedName = UserRoles.Receptionist.ToUpper()
+            },
+            new IdentityRole<Guid>()
```
Remove parentheses or add them in other places.

sdechko1 reviewed (changes requested):
Fix whitespaces where unneccessary or they are doubled.

aleh-kasiankou-innowise committed:
add testing + docker hub deployment workflow

aleh-kasiankou-innowise committed:
Simplify AuthenticationController

-- Move logic to UserManagementService + EmailHandler service

aleh-kasiankou-innowise committed:
move exception handling to middleware & filters

aleh-kasiankou-innowise committed:
fix class naming, remove redundant code, fix formatting

aleh-kasiankou-innowise committed:
move jwt configuration to config files

aleh-kasiankou-innowise committed:
Merge pull request #7 from aleh-kasiankou-innowise/fix/NIS-5220-PR-6

Fix/NIS-5220-PR-6-fixes

aleh-kasiankou-innowise committed:
fix hardcoded jwt validation config values

sdechko1 commented on Innowise.Clinic.Auth.Api/Controllers/ValidationController.cs:
```diff
@@ -0,0 +1,43 @@
+/// </summary>
+[ApiController]
+[Route("[controller]")]
+public class ValidationController : ControllerBase
```
this arguments ([ApiController] and [Route]) can be extracted to base class to reduce duplications like 
```
[ApiController]
[Route("api/controller")]
public abstract ApiController : ControllerBase 
{ }
```

sdechko1 commented on Innowise.Clinic.Auth.Api/Controllers/ValidationController.cs:
```diff
@@ -0,0 +1,43 @@
+    [HttpGet("email/{email:required}")]
+    [ProducesResponseType(typeof(void), 200)]
+    [ProducesResponseType(typeof(void), 400)]
+    public async Task<IActionResult> CheckEmailUniqueness([FromRoute] string email)
```
Why you decided to pass email to query and not into body ?

sdechko1 commented on Innowise.Clinic.Auth.Middleware/AuthenticationExceptionHandlingMiddleware.cs:
```diff
@@ -0,0 +1,53 @@
+        {
+            await next(context);
+        }
+
```
Remove these blank lines

sdechko1 commented on Innowise.Clinic.Auth.Middleware/AuthenticationExceptionHandlingMiddleware.cs:
```diff
@@ -0,0 +1,53 @@
+            context.Response.StatusCode = 500;
+
+            await WriteExceptionMessageToResponse(DefaultUnhandledErrorMessage, context);
+        }
```
Cannot this behavior be achieved by using `return BadRequest(DefaultUnhandledErrorMessage)`?

sdechko1 commented on Innowise.Clinic.Auth.DependencyConfiguration/ConfigurationManager.cs:
```diff
@@ -0,0 +1,113 @@
+
+    private static IdentityBuilder AddPasswordValidators(this IdentityBuilder builder)
+    {
+        builder.AddPasswordValidator<MaximalPasswordLengthValidator<IdentityUser<Guid>>>();
```
Remove this blank line

sdechko1 commented on Innowise.Clinic.Auth.DependencyConfiguration/ConfigurationManager.cs:
```diff
@@ -0,0 +1,113 @@
+        return services;
+    }
+
+    private static IdentityBuilder AddPasswordValidators(this IdentityBuilder builder)
```
Private class members should be declared after public

sdechko1 commented on Innowise.Clinic.Auth.DependencyConfiguration/ConfigurationManager.cs:
```diff
@@ -0,0 +1,113 @@
+
+namespace Innowise.Clinic.Auth.DependencyConfiguration;
+
+public static class ConfigurationManager
```
Since it is not really ConfigurationManager it should be named as ConfigurationExtensions or smth like that.

sdechko1 commented on Innowise.Clinic.Auth.Extensions/AuthTokenPairExtensions.cs:
```diff
@@ -0,0 +1,31 @@
+    {
+        var refreshToken = new JwtSecurityToken(tokens.RefreshToken);
+
+        var tokenIdClaim = refreshToken.Claims.SingleOrDefault(x => x.Type == "jti");
```
In case if in db would be 2 or more tokens this method will throw an Exception so consider using FirstOrDefault for that

aleh-kasiankou-innowise replied:
The database is not used here. The method extracts claim from the refresh token. Valid refresh token can only have one Id claim. So I believe the exception should be thrown if token has more than one claim. It would be a signal that the token is invalid.

sdechko1 commented on Innowise.Clinic.Auth.Extensions/AuthTokenPairExtensions.cs:
```diff
@@ -0,0 +1,31 @@
+
+        var tokenIdClaim = refreshToken.Claims.SingleOrDefault(x => x.Type == "jti");
+
+        if (tokenIdClaim != null) return Guid.Parse(tokenIdClaim.Value);
```
Personally for one line if statements are not really clear and most of the time I saw them as full-filled once

sdechko1 commented on Innowise.Clinic.Auth.IntegrationTesting/IntegrationTestingWebApplicationFactory.cs:
```diff
@@ -0,0 +1,133 @@
+
+    private static int GetFreeTcpPort()
+    {
+        var l = new TcpListener(IPAddress.Loopback, 0);
```
l is not really clear consider name it as listener

sdechko1 commented on Innowise.Clinic.Auth.IntegrationTesting/IntegrationTestingWebApplicationFactory.cs:
```diff
@@ -0,0 +1,133 @@
+public class IntegrationTestingWebApplicationFactory : WebApplicationFactory<Program>, IAsyncLifetime
+{
+    private const string ContainerHost = "localhost";
+
```
Remove this blank line

sdechko1 commented on Innowise.Clinic.Auth.IntegrationTesting/SignInTests.cs:
```diff
@@ -0,0 +1,108 @@
+    {
+        // Arrange
+
+        var validUserRegistrationData = new PatientCredentialsDto
+        {
+            Email = $"test{TestHelper.UniqueNumber}@test.com",
+            Password = "12345678"
+        };
+
+        await _httpClient.PostAsJsonAsync(TestHelper.SignUpEndpointUri, validUserRegistrationData);
```
Why we need this ?

aleh-kasiankou-innowise replied:
As far as I understand we cannot control the order of tests execution, so it would be wrong to use data added to the database during previous tests. That is why each test creates its own patient.

sdechko1 replied:
We can control test cases order but I understand what you are arranging here so It's OK

sdechko1 commented on Innowise.Clinic.Auth.IntegrationTesting/SignInTests.cs:
```diff
@@ -0,0 +1,108 @@
+        // Assert
+
+        Assert.False(response.IsSuccessStatusCode);
+        Assert.Equal("Either an email or a password is incorrect", responseMessage);
```
How email can be incorrect if we are passing exact same value ?

aleh-kasiankou-innowise replied:
Indeed, email cannot be incorrect during this test. However, there is a requirement that states that in case of failed login the following message should be returned:  "Either an email or a password is incorrect".

Here is the link for the requirement list (AC-5) - https://innowise-group.atlassian.net/wiki/spaces/ND/pages/2604499166/US-2+Sign+in

sdechko1 replied:
Strange requirement but what can we do

sdechko1 commented on Innowise.Clinic.Auth.IntegrationTesting/SignUpTests.cs:
```diff
@@ -0,0 +1,152 @@
+            AllowAutoRedirect = false
+        });
+
+        _patientRoleId = _factory.UseDbContext(x => x.Roles.Single(r => r.Name == UserRoles.Patient).Id);
```
For those purposes consider using First()

sdechko1 commented on Innowise.Clinic.Auth.Jwt/Interfaces/ITokenValidator.cs:
```diff
@@ -0,0 +1,10 @@
+
+public interface ITokenValidator
+{
+    Task<ClaimsPrincipal> ValidateTokenPairAndExtractPrincipal(AuthTokenPairDto authTokens,
```
Why validator should exctract Principal ?

aleh-kasiankou-innowise replied:
The principal is extracted to minimize the number of database connections. When I reissue token, I copy the information from a previous one. 

At the moment, the security token lives 15 minutes. So it is indeed possible to call database and extract user principle from there, however, if we decrease token lifetime to 1 minute to enhance security, the number of database requests will skyrocket and the authentication service will immediately become the bottleneck of the system.

sdechko1 commented on Innowise.Clinic.Auth.Jwt/JwtSettings.cs:
```diff
@@ -0,0 +1,14 @@
+    public string ValidIssuer { get; init; }
+
+    public string ValidAudience { get; init; }
+
```
Remove this blank line

sdechko1 commented on Innowise.Clinic.Auth.Jwt/JwtSettings.cs:
```diff
@@ -0,0 +1,14 @@
+
+public record JwtSettings
+{
+    public string Key { get; init; }
```
If you are adding blank lines add here too

sdechko1 commented on Innowise.Clinic.Auth.Jwt/RefreshTokenRevoker.cs:
```diff
@@ -0,0 +1,69 @@
+    public async Task RevokeTokenAsync(Guid tokenId)
+    {
+        var token = await _dbContext.RefreshTokens.FirstOrDefaultAsync(x => x.TokenId == tokenId);
+        if (token == null) throw new UnknownRefreshTokenException();
```
Same as previous comment about one line if-statements

sdechko1 commented on Innowise.Clinic.Auth.Jwt/TokenValidator.cs:
```diff
@@ -0,0 +1,96 @@
+
+            return principal;
+        }
+
```
Remove this blank line

sdechko1 commented on Innowise.Clinic.Auth.UserManagement/Exceptions/EmailConfirmationFailedException.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    public const string DefaultMessage = "The Email confirmation failed. Please try again later.";
+
+    public EmailConfirmationFailedException(IEnumerable<IdentityError> errors) : base(DefaultMessage, StatusCode,
```
if you split arguments by lines consider splitting all of them

sdechko1 commented on Innowise.Clinic.Auth.UserManagement/Exceptions/EmailConfirmationFailedException.cs:
```diff
@@ -0,0 +1,16 @@
+    {
+    }
+
+    public new static int StatusCode => 400;
```
Why do you need new here ?

aleh-kasiankou-innowise committed:
Move services to Services project

aleh-kasiankou-innowise committed:
add integration with profiles service

aleh-kasiankou-innowise committed:
add integration with profile services

aleh-kasiankou-innowise committed:
fix integration tests

-- created a distinct method for registration and token retrieval to reuse the code in a number of tests

-- created TokenNotReturned Exception that will be thrown during tests if token is not returned after login

aleh-kasiankou-innowise committed:
improve exception handling

aleh-kasiankou-innowise committed:
separate validation configurations

aleh-kasiankou-innowise committed:
Merge pull request #8 from aleh-kasiankou-innowise/feature/compatibility-with-other-services

compatibility with other services + code refactoring

aleh-kasiankou-innowise committed:
create base class for all controllers

aleh-kasiankou-innowise committed:
move user tokens check to internal controller

aleh-kasiankou-innowise committed:
validate emails via post requests

aleh-kasiankou-innowise committed:
fix newlines

aleh-kasiankou-innowise committed:
remove inline if statements

aleh-kasiankou-innowise committed:
remove property hiding from exceptions

aleh-kasiankou-innowise committed:
Merge pull request #9 from aleh-kasiankou-innowise/fix/NIS-5220-PR-6

Fix/NIS-5220-PR-6

aleh-kasiankou-innowise committed:
fix variable naming

aleh-kasiankou-innowise committed:
Merge pull request #10 from aleh-kasiankou-innowise/fix/NIS-5220-PR-6

fix variable naming

aleh-kasiankou-innowise committed:
fix parameter + method naming & code formatting

aleh-kasiankou-innowise committed:
rename configuration manager -> configuration extensions

aleh-kasiankou-innowise committed:
Merge pull request #11 from aleh-kasiankou-innowise/fix/NIS-5220-PR-6

Fix/NIS-5220-PR-6
